checks/permissions.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020 Security Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checks
+
+import (
+	"fmt"
+
+	"github.com/ossf/scorecard/checker"
+	"gopkg.in/yaml.v2"
+)
+
+const CheckPermissions = "Token-Permissions"
+
+//nolint:gochecknoinits
+func init() {
+	registerCheck(CheckPermissions, leastPrivilegedTokens)
+}
+
+func leastPrivilegedTokens(c *checker.CheckRequest) checker.CheckResult {
+	return CheckFilesContent(CheckPermissions, ".github/workflows/*", true, c, validateGitHubActionTokenPermissions)
+}
+
+type GitHubActionWorkflowConfig struct {
+	Permissions *struct {
+		Contents       *string `yaml:"contents"`
+		Actions        *string `yaml:"actions"`
+		Checks         *string `yaml:"checks"`
+		Deploymments   *string `yaml:"deployments"`
+		Issues         *string `yaml:"issues"`
+		Metadata       *string `yaml:"metadata"`
+		Packages       *string `yaml:"packages"`
+		PullRequests   *string `yaml:"pull requests"`
+		Projects       *string `yaml:"repository projects"`
+		SecurityEvents *string `yaml:"security events"`
+		Statuses       *string `yaml:"statuses"`
+	} `yaml:"permissions"`
+	Name string `yaml:"name"`
+}
+
+func isPermissionWrite(permission *string) bool {
+	return permission != nil &&
+		*permission != "read" &&
+		*permission != "none"
+}
+
+func permissionToString(permission *string) string {
+	if permission != nil {
+		return *permission
+	}
+	return "read"
+}
+
+func validateDefaultReadPermissions(config GitHubActionWorkflowConfig, path string,
+	logf func(s string, f ...interface{})) bool {
+	if config.Permissions == nil {
+		logf("!! token-permissions/token - no permission defined in %v", path)
+		return false
+	}
+
+	r := true
+	if isPermissionWrite(config.Permissions.Contents) {
```

Comment:
as above let's deserialize the permissions field into a `map[string]string` here instead and just iterate here.